internal/api/v1/application/stage.go

```diff
@@ -339,50 +340,153 @@ func validateBlob(ctx context.Context, blobUID string, app models.AppRef, s3Conn
 	return nil
 }
 
-// newPipelineRun is a helper which creates a Tekton pipeline run
-// resource from the given staging params
-func newPipelineRun(app stageParam) *v1beta1.PipelineRun {
-	str := v1beta1.NewArrayOrString
+// newJobRun is a helper which creates the Job related resources from
+// the given staging params. That is the job itself, and a secret
+// holding the job's environment. Which is a copy of the app
+// environment + standard variables.
+func newJobRun(app stageParam) (*batchv1.Job, *corev1.Secret) {
+
+	// fake stage params of the previous to pull the old image url from.
+	previous := app
+	previous.Stage = models.NewStage(app.PreviousStageID)
 
 	protocol := "http"
 	if app.S3ConnectionDetails.UseSSL {
 		protocol = "https"
 	}
-	awsScript := fmt.Sprintf("aws --endpoint-url %s://$1 s3 cp s3://$2/$3 $(workspaces.source.path)/$3", protocol)
-	awsArgs := []string{
-		app.S3ConnectionDetails.Endpoint,
-		app.S3ConnectionDetails.Bucket,
-		app.BlobUID,
+
+	awsScript := fmt.Sprintf("echo Extracting _ _ __ ___ _____ $(whoami) ; aws --endpoint-url %s://%s s3 cp s3://%s/%s /workspace/source/%s ; echo _ _ __ ___ _____ Extracted",
+		protocol, app.S3ConnectionDetails.Endpoint, app.S3ConnectionDetails.Bucket,
+		app.BlobUID, app.BlobUID)
+
+	unpackScript := fmt.Sprintf("echo Unpacking _ _ __ ___ _____ $(whoami) ; mkdir /workspace/source/app ; tar -xvf /workspace/source/%s -C /workspace/source/app/; rm /workspace/source/%s; chown -R 1000:1000 /workspace ; cp -vL /platform/appenv/* /platform/env/ ; ls -la /platform/env ; echo _ _ __ ___ _____ Unpacked", app.BlobUID, app.BlobUID)
+
+	buildpackScript := fmt.Sprintf("echo Creating _ _ __ ___ _____ $(whoami) ; ls -la /platform/env ; /cnb/lifecycle/creator -app=/workspace/source/app -cache-dir=/workspace/cache -uid=1000 -gid=1000 -layers=/layers -platform=/platform -report=/layers/report.toml -process-type=web -skip-restore=false -previous-image=%s %s && curl -X POST http://localhost:4191/shutdown",
```

Comment:
This line is assuming linkerd is installed: `curl -X POST http://localhost:4191/shutdown`.
When linkerd it not there, it fails (see my other comment too). I don't think the staging script need to know anything about linkerd. If I understand correctly this is a problem of linkerd not playing nice with jobs right? So we are using this to work around the issue: https://github.com/linkerd/linkerd2-proxy/pull/811

For now, I would suggest we make the request but not fail if it's not there to avoid extending this PR more. I need to look it up again but I think we only need to mesh the staging Pod is because there was not way to trust the registry certificate when it was self signed. Now that we have more control over the staging process, maybe we can fix the original problem and avoid the internal registry url hack along with linkerd on the staging namespace. But let's do that outside this PR.

Reply:
Reading [here](https://github.com/buildpacks/pack/issues/159#issuecomment-563361568): `but you can workaround that by adding certs to your system` . So maybe all we need to do, it to mount the registry certificate inside the staging pod like we did here: https://github.com/epinio/helm-charts/blob/d1d8e97f49b5e103556aea22fc5d81be543abd05/chart/epinio/templates/tekton-buildpacks-task.yaml#L159). I wonder, is this what [the line below does](https://github.com/epinio/epinio/pull/1141/files#diff-1d5683322c8ac645535eab16a4aa0ef4a516edecf543777e7de0b8ca5f4f7c7fR385)? (`/home/cnb/.docker/`) If yes, then we don't need to mesh the staging pod and we don't need the internal registry url anymore.

We still need to encrypt the communication with S3 but we don't need to worry about that, because [S3 is going away soon](#1157 ).

Reply:
Sorry wrong line, I meant `registry-cert` not `registry-creds`. Here is the relevant code: https://github.com/epinio/epinio/pull/1141/files#diff-1d5683322c8ac645535eab16a4aa0ef4a516edecf543777e7de0b8ca5f4f7c7fR463-R476

Reply:
ok I found the description of the situation: https://github.com/epinio/epinio/blob/70075c77758c4ff9bb26a9caff0e8b8cb5971faa/internal/api/v1/application/deploy.go#L282-L304
So we still need the localhost trick (that's for kube not tekton/staging :facepalm: ). For staging we access the registry through the ingress and we mount the cert. 

So the only reason to have the staging namespace meshed, is S3. Given we will remove that, I would suggest we  keep the `curl` here but make sure it doesn't fail when linkerd is not installed. We will remove the `curl` when we remove S3 and then we will also "unmesh" the staging namespace. If we go this way, let's put a TODO on the #1157 issue to remember that.

Reply:
Heh, Yes, just proposed that in the slack, `curl || true`.
Yes please on the additional work wrt linkerd and hacks.

Reply:
Done the `|| true`

Reply:
Added todo to #1157